rust/connlib/tunnel/src/lib.rs

```diff
@@ -172,8 +194,69 @@ impl<CB> Tunnel<CB, GatewayState>
 where
     CB: Callbacks + 'static,
 {
-    pub fn poll_next_event(&self, cx: &mut Context<'_>) -> Poll<Event<ClientId>> {
-        self.poll_next_event_common(cx)
+    pub fn poll_next_event(&self, cx: &mut Context<'_>) -> Poll<Result<Event<ClientId>>> {
+        loop {
+            {
+                let mut device = self.device.write();
+
+                match device.as_mut().map(|d| d.poll_read(cx)) {
+                    Some(Poll::Ready(Ok(Some(packet)))) => {
+                        let dest = packet.destination();
+
+                        let peers_by_ip = self.peers_by_ip.read();
+                        let Some(peer) = peer_by_ip(&peers_by_ip, dest) else {
+                            continue;
+                        };
+                        let client_id = peer.inner.conn_id;
+                        let mut buf = self.buf.lock();
+
+                        match peer.inner.encapsulate(packet, dest, buf.as_mut_slice()) {
+                            Ok(None) => {}
+                            Ok(Some(b)) => {
+                                tokio::spawn({
+                                    let channel = peer.channel.clone();
+                                    let mut sender = self.stop_peer_command_sender.clone();
+
+                                    async move {
+                                        if let Err(e) = channel.write(&b).await {
+                                            tracing::error!(resource_address = %dest, err = ?e, "failed to handle packet {e:#}");
+                                            let _ = sender.send(client_id).await;
+                                        }
+                                    }
+                                });
+                            }
+                            Err(e) => {
+                                tracing::error!(resource_address = %dest, err = ?e, "failed to handle packet {e:#}");
+
+                                if e.is_fatal_connection_error() {
+                                    self.peers_to_stop.lock().push_back(client_id);
+                                }
+                            }
+                        };
+
+                        continue;
+                    }
+                    Some(Poll::Ready(Ok(None))) => {
+                        tracing::info!("Device stopped");
+                        drop(device.take());
```

Comment:
I think we don't need to explicitly `drop`

Reply:
We don't but just calling `device.take()` looks weird and probably needs a comment. Instead of writing a comment, I figured calling `drop` explicitly makes it obvious that we want to de-allocate the device.